app/app_aes.c

```diff
@@ -23,6 +27,642 @@ void app_aes_cleanup(void) {
     glb_cipher_ctx = NULL;
 }
 
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+
+int app_aes_handler(ACVP_TEST_CASE *test_case) {
+    ACVP_SYM_CIPHER_TC  *tc = NULL;
+    EVP_CIPHER_CTX *cipher_ctx = NULL;
+    EVP_CIPHER *cipher = NULL;
+    const char *type = NULL, *mode = NULL;
+    char alg_name[32];
+    unsigned char *iv = NULL;
+    int rv = 1, direction = 0;
+    ACVP_SUB_AES alg;
+    OSSL_PARAM_BLD *pbld = NULL;
+    OSSL_PARAM *params = NULL;
+
+    if (!test_case) {
+        return rv;
+    }
+
+    tc = test_case->tc.symmetric;
+
+    if (glb_cipher_ctx == NULL) {
+        glb_cipher_ctx = EVP_CIPHER_CTX_new();
+        if (glb_cipher_ctx == NULL) {
+            printf("Failed to allocate global cipher_ctx");
+            return 1;
+        }
+    }
+
+    /* Begin encrypt code section */
+    cipher_ctx = glb_cipher_ctx;
+    if ((tc->test_type != ACVP_SYM_TEST_TYPE_MCT)) {
+        EVP_CIPHER_CTX_init(cipher_ctx);
+    }
+
+    alg = acvp_get_aes_alg(tc->cipher);
+    if (alg == 0) {
+        printf("Invalid cipher value");
+        return 1;
+    }
+
+    switch (tc->key_len) {
+    case 128:
+        type = "AES-128-";
+        break;
+    case 192:
+        type = "AES-192-";
+        break;
+    case 256:
+        type = "AES-256-";
+        break;
+    default:
+        printf("Unsupported AES key length\n");
+        rv = 1;
+        goto err;
+    }
+
+    switch (alg) {
+    case ACVP_SUB_AES_ECB:
+        mode = "ECB";
+        break;
+    case ACVP_SUB_AES_CTR:
+        mode = "CTR";
+        iv = tc->iv;
+        break;
+    case ACVP_SUB_AES_CFB1:
+        mode = "CFB1";
+        iv = tc->iv;
+        break;
+    case ACVP_SUB_AES_CFB8:
+        mode = "CFB8";
+        iv = tc->iv;
+        break;
+    case ACVP_SUB_AES_CFB128:
+        mode = "CFB";
+        iv = tc->iv;
+        break;
+    case ACVP_SUB_AES_OFB:
+        mode = "OFB";
+        iv = tc->iv;
+        break;
+    case ACVP_SUB_AES_CBC:
+        mode = "CBC";
+        iv = tc->iv;
+        break;
+    case ACVP_SUB_AES_CBC_CS1:
+    case ACVP_SUB_AES_CBC_CS2:
+    case ACVP_SUB_AES_CBC_CS3:
+        mode = "CBC-CTS";
+        iv = tc->iv;
+        break;
+    case ACVP_SUB_AES_XTS:
+        mode = "XTS";
+        if (tc->tw_mode == ACVP_SYM_CIPH_TWEAK_HEX) {
+            iv = tc->iv;
+        }
+        break;
+    case ACVP_SUB_AES_GCM:
+    case ACVP_SUB_AES_GCM_SIV:
+    case ACVP_SUB_AES_CCM:
+    case ACVP_SUB_AES_XPN:
+    case ACVP_SUB_AES_KW:
+    case ACVP_SUB_AES_KWP:
+    case ACVP_SUB_AES_GMAC:
+    default:
+        printf("Error: Unsupported AES mode requested by ACVP server\n");
+        rv = 1;
+        goto err;
+    }
+
+    memzero_s(alg_name, 32);
+    strncat_s(alg_name, 32, type, strnlen_s(type, 16));
+    strncat_s(alg_name, 32, mode, strnlen_s(mode, 16));
```

Comment:
does mode overwrite type ?

Reply:
strncat works by looking for a null terminator before the end of a given buffer and concatenating starting there; so the memzero prevents that.